libs/constrained-generators/src/Constrained/Test.hs

```diff
@@ -0,0 +1,552 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Constrained.Test where
+
+import Data.Int
+import Data.Map (Map)
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Typeable
+import Data.Word
+import GHC.Generics
+import GHC.Natural
+import Test.QuickCheck hiding (Args, Fun, forAll)
+import Test.QuickCheck qualified as QC
+import Test.Tasty
+import Test.Tasty.QuickCheck (testProperty)
+
+-- import Constrained.Base
+-- import Constrained.Core
+-- import Constrained.Env
+-- import Constrained.GenT
+-- import Constrained.Instances ()
+-- import Constrained.Spec
+
+-- import Constrained.Univ
+
+import Constrained
+
+-- Properties -------------------------------------------------------------
+
+prop_conformEmpty ::
+  forall fn a.
+  HasSpec fn a =>
+  a ->
+  Property
+prop_conformEmpty a = property $ conformsTo @fn a (emptySpec @fn @a)
+
+-- prop_propagateSpec ::
+--   (HasSpec fn a, HasSpec fn b) =>
+--   Var b ->
+--   Term fn a ->
+--   Spec fn a ->
+--   Property
+-- prop_propagateSpec var tm spec =
+--   let spec' = errorGE $ do
+--         ctx <- toCtx var tm
+--         pure $ propagateSpec spec ctx
+--    in QC.forAll (strictGen $ genFromSpec spec') $ \geval -> fromGE (\err -> counterexample (unlines err) False) $ do
+--         val <- geval
+--         res <- runTerm (singletonEnv var val) tm
+--         pure $ property $ conformsToSpec res spec
+
+-- prop_mapSpec ::
```

Comment:
If we eventually want to uncomment these. Add some TODO comments. If not we should consider deleting them.